### ISSUE TITLE
"'Future' object is not subscriptable" after persisting an object with `dask.distributed`

### ISSUE THREAD

**Issue by bgbg** [dask/dask#4361]
There is a piece of code that runs with the single-machine scheduler but not when `dask.distributed` is used. The code below adds many computed columns to a dataframe. To prevent RecursionError (see #4360), I call `persist` once in a while

```
import dask.dataframe as dd
import pandas as pd
import random
import itertools
import string
import dask.dataframe as dd
random.seed(10)

bigrams = [''.join(b) for b in itertools.combinations_with_replacement(string.ascii_letters, 2)]
data = dd.from_pandas(
    pd.DataFrame(
        {
            'word': [''.join(random.choices(string.ascii_letters,k=4)) for _ in range(100)]
        }
    ), 
    npartitions=8
    )

count = 0
K = 500
for bigram in tqdm.tqdm_notebook(bigrams):
    col_name = f'has_bigram_{bigram}'
    data[col_name] = data['word'].apply(lambda w: bigram in w, meta=(col_name, bool))
    count += 1
    if count > K:
        data = data.persist()
        count = 0
if count:
    data.persist()

data.compute()
```

Everything works OK. However, if I add the following lines to the code above, I get `TypeError: 'Future' object is not subscriptable`. :

```from dask.distributed import Client
client = Client()
```

**Comment by mrocklin** [comment]
Thank you for the issue @bgbg .  Can I ask you to try to reduce the example here?  The smaller you can make this the easier it is to convince maintainers to try it out.  So for example do you need to run `tqdm` in order to reproduce this error?  Do you need to have 600 columns?  Do you need to run persist?  etc.  You're probably the person who is most capable at reducing the problem while still reproducing the error.  See http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports

**Comment by bnaul** [comment]
I have also run into this and have random .persist() calls sprinkled around a couple of notebooks that seemed to resolve the issue; I think it's related to calling .apply() but never came up with a simple example. If I run into it again I'll try to look into it.

**Comment by bgbg** [comment]
> So for example do you need to run tqdm in order to reproduce this error? 
> Do you need to have 600 columns? 
> Do you need to run persist? etc. 

Yes. I suspect that the problem is caused by the the combination of adding a large # of columns and calling `persist`. Reducing the # of columns solves the problem.

If you add this many columns and don't call persist, you get another problem (#4360).


You were right about tqdm, though . Following is the snippet without it:


```
import dask.dataframe as dd
import pandas as pd
import random
import itertools
import string
import dask.dataframe as dd
random.seed(10)

bigrams = [''.join(b) for b in itertools.combinations_with_replacement(string.ascii_letters, 2)]
data = dd.from_pandas(
    pd.DataFrame(
        {
            'word': [''.join(random.choices(string.ascii_letters,k=4)) for _ in range(100)]
        }
    ), 
    npartitions=8
    )

count = 0
K = 50
for bigram in bigrams:
    col_name = f'has_bigram_{bigram}'
    data[col_name] = data['word'].apply(lambda w: bigram in w, meta=(col_name, bool))
    count += 1
    if count > K:
        data = data.persist()
        count = 0
if count:
    data.persist()

data.compute()
```

**Comment by mrocklin** [comment]
If you get a chance, reading through and then implementing the suggestions in http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports would be helpful .

I also suspect that this would be resolved by the work referred to in https://github.com/dask/dask/pull/4229/files#r242719752 cc @jcrist

**Comment by bgbg** [comment]
@mrocklin 
Thank you for pointing me to the guideline. However, I fail to understand how my code breaks those guidelines, except for the forgotten reference to tqdm.

As far as I can see, the snippet that I provided is verifiable and complete. It is somewhat long but it is, in my opinion, the minimal version that is needed to demonstrate the problem I reported here.

**Comment by mrocklin** [comment]
- Does this only happen with bigram computations?
- How about strings?  Does it happen with integers too?
- Is the dataframe assignment necessary?
-  ...

OK.  Regardless, thanks for your effort here in producing the example you have so far.  Hopefully one of the maintainers has time to investigate this code.  I probably won't get to it near-term but maybe someone else will have time?  cc @martindurant @TomAugspurger @jcrist .  I agree with you @bgbg that this bug seems significant.

**Comment by TomAugspurger** [comment]
Will try to take a look later today.

**Comment by TomAugspurger** [comment]
I wasn't able to reproduce with the following

```python
from distributed import Client
import dask.dataframe as dd
import pandas as pd
import random
import itertools
import string
import dask.dataframe as dd
random.seed(10)


def main():
    client = Client()

    bigrams = [''.join(b) for b in itertools.combinations_with_replacement(string.ascii_letters, 2)]
    data = dd.from_pandas(
        pd.DataFrame(
            {
                'word': [''.join(random.choices(string.ascii_letters,k=4)) for _ in range(100)]
            }
        ),
        npartitions=8
    )

    count = 0
    K = 50
    for bigram in bigrams:
        col_name = f'has_bigram_{bigram}'
        data[col_name] = data['word'].apply(lambda w: bigram in w, meta=(col_name, bool))
        count += 1
        if count > K:
            data = data.persist()
            count = 0

    if count:
        data.persist()

    data.compute()


if __name__ == '__main__':
    main()
```

@bgbg what versions of dask, distributed, and Python are you using?

**Comment by bnaul** [comment]
@TomAugspurger running your above code on MacOS, Python 3.7.1, dask==1.0.0, distributed==1.25.2, I see:
```
distributed.worker - WARNING -  Compute Failed
Function:  subgraph_callable
args:      ()
kwargs:    {}
Exception: TypeError("'Future' object is not subscriptable")

distributed.worker - WARNING -  Compute Failed
Function:  subgraph_callable
args:      ()
kwargs:    {}
Exception: TypeError("'Future' object is not subscriptable")

...
```

**Comment by bgbg** [comment]
```
print(sys.version)
print(dask.__version__)
print(distributed.__version__)
```

Gives
```
3.6.7 | packaged by conda-forge | (default, Nov 21 2018, 02:32:25) 
[GCC 4.8.2 20140120 (Red Hat 4.8.2-15)]
1.0.0
1.25.1
```

I have just checked the same code on another machine (a Mac), and everything worked there:

```
3.6.7 | packaged by conda-forge | (default, Nov 20 2018, 18:20:05) 
[GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.37)]
0.17.5
1.21.8
```

**Comment by gsakkis** [comment]
Probably the same with https://github.com/dask/distributed/issues/2460

**Comment by Mdhvince** [comment]
@TomAugspurger  I have the same problem, how can I try dask master on my problem ?

**Comment by TomAugspurger** [comment]
There may have been two issues here, one of which is not yet resolved. I'd
recommend subscribing to https://github.com/dask/distributed/issues/2460

**Comment by jcrist** [comment]
This seems to have been resolved by #2505:

```python
import dask
import distributed
import dask.dataframe as dd
import pandas as pd
import random
import itertools
import string
import dask.dataframe as dd

print("Dask version: %s" % dask.__version__)
print("Distributed version: %s" % distributed.__version__)

random.seed(10)

bigrams = [''.join(b) for b in itertools.combinations_with_replacement(string.ascii_letters, 2)]
data = dd.from_pandas(
    pd.DataFrame(
        {
            'word': [''.join(random.choices(string.ascii_letters,k=4)) for _ in range(100)]
        }
    ),
    npartitions=8
)

count = 0
K = 500
for bigram in bigrams:
    col_name = f'has_bigram_{bigram}'
    data[col_name] = data['word'].apply(lambda w: bigram in w, meta=(col_name, bool))
    count += 1
    if count > K:
        data = data.persist()
        count = 0
if count:
    data.persist()

data.compute()
print("OK!")
```

```bash
$ python test.py
Dask version: 1.1.5+7.g3205dbd
Distributed version: 1.26.1+1.g50dd21f
OK!
```

Things still aren't as optimal as they could be (higher level fusion as described in #4360 would help), but this specific issue is resolved. Closing.
